File: backend/schemas/invoice.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

def _trim_invoice_number(v: str | None) -> str | None:
    if v is None:
        return None
    text = str(v).strip()
    return text if text else None


def _clean_paid_by(v: str | None) -> str | None:
    if v is None:
        return None
    text = str(v).strip()
    if not text:
        return None
    return text[:300]


class ExtractionResult(BaseModel):
    invoice_date: str | None = None
    name_of_company: str | None = None
    address_of_company: str | None = None
    invoice_number: str | None = None

    @field_validator("invoice_number", mode="before")
    @classmethod
    def _invoice_number_format(cls, v: str | None) -> str | None:
        return _trim_invoice_number(v)
    amount: float | None = None
    debt: float | None = None
    currency: str | None = None
    # Set by Vision for utility routing; not stored on Invoice row
    document_type: str | None = None
    account_details: str | None = None
    internal_note_description: str | None = None
    client_employee_related: str | None = None
    category: str | None = None
    confidence_score: float = 0.0
    needs_review: bool = False
    # Per-field confidence scores, e.g. {"name_of_company": 0.95, "amount": 0.72}
    field_confidences: dict[str, float] | None = None
    review_reasons: list[str] = Field(default_factory=list)


class InvoiceUpdate(BaseModel):
    invoice_date: date | None = None
    name_of_company: str | None = None
    address_of_company: str | None = None
    invoice_number: str | None = None

    @field_validator("invoice_number", mode="before")
    @classmethod
    def _invoice_number_format(cls, v: str | None) -> str | None:
        return _trim_invoice_number(v)
    amount: Decimal | None = None
    debt: Decimal | None = None
    currency: str | None = None
    original_amount: Decimal | None = None
    original_currency: str | None = None
    account_details: str | None = None
    internal_note_description: str | None = None
    client_employee_related: str | None = None
    paid_by: str | None = None
    fixed_status: str | None = None
    category: str | None = None

    @field_validator("paid_by", mode="before")
    @classmethod
    def _paid_by_format(cls, v: str | None) -> str | None:
        return _clean_paid_by(v)


class InvoiceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    invoice_date: date | None
    name_of_company: str | None
    address_of_company: str | None
    invoice_number: str | None
    invoice_number_normalized: str | None = None
    amount: Decimal | None
    debt: Decimal | None
    currency: str | None
    original_amount: Decimal | None
    original_currency: str | None
    exchange_rate: Decimal | None
    exchange_rate_date: date | None
    account_details: str | None
    internal_note_description: str | None
    client_employee_related: str | None
    paid_at_date: date | None
    paid_by: str | None
    fixed_status: str | None
    category: str | None
    extraction_confidence: Decimal | None
    field_confidences: dict | None
    review_reasons: list[str] | None = None
    review_status: str
    match_status: str
    uploaded_by: int
    source_file_id: int | None
    source_filename: str | None = None
    source_mime_type: str | None = None
    upload_source: str | None = None
    ingest_sender_email: str | None = None
    ingest_sender_name: str | None = None
    ingest_email_subject: str | None = None
    ingest_message_id: str | None = None
    created_at: datetime
    updated_at: datetime


class InvoiceListResponse(BaseModel):
    items: list[InvoiceResponse]
    total: int
    page: int
    limit: int


class UploadItemResponse(BaseModel):
    upload_id: int
    original_filename: str
    processing_status: str
    invoice_id: int | None = None
    error: str | None = None
    # Set when processing_status is "linked" (duplicate file already in system).
    message: str | None = None
    original_uploader_email: str | None = None


class InvoiceUploadResponse(BaseModel):
    uploaded: int
    items: list[UploadItemResponse]


class InvoiceApproveResponse(BaseModel):
    id: int
    review_status: str

File: frontend/src/types/invoice.ts
/** Phase 3: Documents drawer Save sends partial Invoice (writable fields only). */

export type ReviewStatus = "pending" | "approved" | "needs_review" | "manual_review";
export type MatchStatus = "unmatched" | "matched" | "partially_matched" | "needs_review";

export interface Invoice {
  id: number;
  invoice_date: string | null;
  name_of_company: string | null;
  address_of_company: string | null;
  invoice_number: string | null;
  /** Canonical matching key; may differ from display (e.g. slashes/hyphens stripped). */
  invoice_number_normalized: string | null;
  amount: number | null;
  debt: number | null;
  currency: string | null;
  original_amount: number | null;
  original_currency: string | null;
  exchange_rate: number | null;
  exchange_rate_date: string | null;
  account_details: string | null;
  internal_note_description: string | null;
  client_employee_related: string | null;
  paid_at_date: string | null;
  paid_by: string | null;
  fixed_status: string | null;
  category: string | null;
  extraction_confidence: number | null;
  /** Per-field confidence map, e.g. { name_of_company: 0.95, amount: 0.72 } */
  field_confidences: Record<string, number> | null;
  /** Why this invoice needs human review (extraction or matching). */
  review_reasons: string[] | null;
  review_status: ReviewStatus;
  match_status: MatchStatus;
  uploaded_by: number;
  source_file_id: number | null;
  source_filename: string | null;
  source_mime_type: string | null;
  upload_source?: string | null;
  ingest_sender_email?: string | null;
  ingest_sender_name?: string | null;
  ingest_email_subject?: string | null;
  ingest_message_id?: string | null;
  created_at: string;
  updated_at: string;
}

export interface InvoiceListResponse {
  items: Invoice[];
  total: number;
  page: number;
  limit: number;
}

export interface InvoiceFilters {
  review_status?: string;
  match_status?: string;
  invoice_date_from?: string;
  invoice_date_to?: string;
  company?: string;
  search?: string;
  sort?: string;
  upload_source?: string;
  page?: number;
  limit?: number;
}

export interface UploadItem {
  upload_id: number;
  original_filename: string;
  processing_status: string;
  invoice_id?: number | null;
  error?: string | null;
  message?: string | null;
  original_uploader_email?: string | null;
}

export interface UploadResponse {
  uploaded: number;
  items: UploadItem[];
}
